src/Ethereum/ContractCall.cpp

```diff
@@ -0,0 +1,109 @@
+// Copyright © 2017-2020 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "ContractCall.h"
+#include "ABI.h"
+#include "HexCoding.h"
+#include "uint256.h"
+#include <boost/algorithm/string/predicate.hpp>
+#include <iostream>
+
+using namespace TW;
+using namespace TW::Ethereum::ABI;
+using namespace std;
+
+using json = nlohmann::json;
+
+static void fillArray(Function& func, const string& type) {
+    auto param = make_shared<ParamArray>();
+    auto baseType = string(type.begin(), type.end() - 2);
+    auto value = ParamFactory::make(baseType);
+    param->addParam(value);
+    func.addParam(param, false);
+}
+
+static void fill(Function& func, const string& type) {
+    if (boost::algorithm::ends_with(type, "[]")) {
+        return fillArray(func, type);
+    } else {
+        auto param = ParamFactory::make(type);
+        func.addParam(param, false);
+    }
+}
+
+static vector<string> getArrayValue(Function& func, const string& type, int idx) {
+    auto baseType = string(type.begin(), type.end() - 2);
+    shared_ptr<ParamBase> param;
+    func.getInParam(idx, param);
+    auto array = dynamic_pointer_cast<ParamArray>(param);
+    auto count = array->getCount();
+    auto result = vector<string>();
+    for (int i = 0; i < count; i++) {
+        result.push_back(ParamFactory::getValue(array->getParam(i), baseType));
+    }
+    return result;
+}
+
+static string getValue(Function& func, const string& type, int idx) {
+    shared_ptr<ParamBase> param;
+    func.getInParam(idx, param);
+    return ParamFactory::getValue(param, type);
+}
+
+static json buildInputs(Function& func, const json& registry) {
+    auto inputs = json::array();
+    for (int i = 0; i < registry["inputs"].size(); i++) {
```

Comment:
Error handling for missing "inputs"?

Reply:
we won't decode arbitrary abi json tho, need to pre calculate method id + correct abi json